[Curl] Add Timing-Allow-Origin checks for Curl port

kshukuwa opened the pull request:
#### 3f6d008b5fa784cb5550442fce0dc0f5240e0463
<pre>
[Curl] Add Timing-Allow-Origin checks for Curl port
<a href="https://bugs.webkit.org/show_bug.cgi?id=247048">https://bugs.webkit.org/show_bug.cgi?id=247048</a>

Reviewed by Fujii Hironori.

Add missing TAO checks to Curl backend.

* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/network/curl/CurlResourceHandleDelegate.cpp:
(WebCore::CurlResourceHandleDelegate::curlDidReceiveResponse):
(WebCore::CurlResourceHandleDelegate::curlDidComplete):
(WebCore::CurlResourceHandleDelegate::updateNetworkLoadMetrics):
* Source/WebCore/platform/network/curl/CurlResourceHandleDelegate.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::create):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::NetworkDataTaskCurl):
(WebKit::NetworkDataTaskCurl::curlDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::curlDidComplete):
(WebKit::NetworkDataTaskCurl::updateNetworkLoadMetrics):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:

Canonical link: <a href="https://commits.webkit.org/256044@main">https://commits.webkit.org/256044@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/95d694988cd30ea91392ebaa639fc004f02dbbad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94314 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3491 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27249 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103981 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164255 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98312 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3534 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31698 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86648 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99989 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99984 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2534 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80708 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29570 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84453 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72467 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38093 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/17914 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35971 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19186 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4177 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39857 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41778 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41806 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38415 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

fujii commented on LayoutTests/platform/wincairo/TestExpectations:
```diff
@@ -2134,3 +2134,7 @@ fast/text/crash-complex-text-surrogate.html [ Failure ]
+
+imported/w3c/web-platform-tests/resource-timing/cross-origin-redirects.html [ Pass ]
+imported/w3c/web-platform-tests/resource-timing/cross-origin-start-end-time-with-redirects.html [ Pass ]
+imported/w3c/web-platform-tests/resource-timing/resource_reuse.sub.html [ Pass ]
```
Does WinCairo WK1 pass these tests? WinCairo buildbot is still running WK1 layout tests.

kshukuwa replied:
Yes. I modified CurlResourceHandleDelegate.cpp that is for WK1 to pass these tests.

fujii commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
```diff
@@ -222,6 +226,8 @@ void NetworkDataTaskCurl::curlDidComplete(CurlRequest&, NetworkLoadMetrics&& net
         return;
     }
 
+    updateNetworkLoadMetrics(networkLoadMetrics);
```
Cocoa port is calling checkTAO in willPerformHTTPRedirection and didReceiveResponse.
On the other hand, NetworkDataTaskCurl::updateNetworkLoadMetrics is called in curlDidComplete and curlDidReceiveResponse.
Is this right?

kshukuwa replied:
This is right. Cocoa port store NetworkLoadMetrics in a member variable of the NetworkDataTaskCocoa class.

https://github.com/WebKit/WebKit/blob/5ab8b234cd6549d90d5d7853b340fbcb3646af97/Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.h#L115

On the other hand, Curl port does not store it and a new object is spawned every time when called CurlHandle::getNetworkLoadMetrics.

https://github.com/WebKit/WebKit/blob/5ab8b234cd6549d90d5d7853b340fbcb3646af97/Source/WebCore/platform/network/curl/CurlContext.cpp#L820

https://github.com/WebKit/WebKit/blob/5ab8b234cd6549d90d5d7853b340fbcb3646af97/Source/WebCore/platform/network/curl/CurlRequest.cpp#L387

https://github.com/WebKit/WebKit/blob/5ab8b234cd6549d90d5d7853b340fbcb3646af97/Source/WebCore/platform/network/curl/CurlRequest.cpp#L482

The generated objects are then notified to curlDidReceiveResponse and curlDidComplete. Therefore, NetworkDataTaskCurl::updateNetworkLoadMetrics must be called in both curlDidReceiveResponse and curlDidComplete to update failsTAOCheck in NetworkLoadMetrics.

fujii commented on Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:
```diff
@@ -93,24 +94,28 @@ class NetworkDataTaskCurl final : public NetworkDataTask, public WebCore::CurlRe
     bool m_blockingCookies { false };
 
-    WebCore::ShouldRelaxThirdPartyCookieBlocking m_shouldRelaxThirdPartyCookieBlocking { WebCore::ShouldRelaxThirdPartyCookieBlocking::No };
+    bool m_failsTAOCheck { false };
```
I'm wondering why only curl port adds `m_failsTAOCheck`? Cocoa and Soup ports are using `m_networkLoadMetrics.failsTAOCheck`.

kshukuwa replied:
The answer is similar to the previous question. Curl port does not store NetworkLoadMetrics like Cocoa and SOUP. Therefor, curl port adds `m_failsTAOCheck` to store results of failsTAOCheck.

fujii reviewed (approved):
LGTM

kshukuwa committed:
[Curl] Add Timing-Allow-Origin checks for Curl port
https://bugs.webkit.org/show_bug.cgi?id=247048

Reviewed by Fujii Hironori.

Add missing TAO checks to Curl backend.

* LayoutTests/platform/wincairo/TestExpectations:
* Source/WebCore/platform/network/curl/CurlResourceHandleDelegate.cpp:
(WebCore::CurlResourceHandleDelegate::curlDidReceiveResponse):
(WebCore::CurlResourceHandleDelegate::curlDidComplete):
(WebCore::CurlResourceHandleDelegate::updateNetworkLoadMetrics):
* Source/WebCore/platform/network/curl/CurlResourceHandleDelegate.h:
* Source/WebKit/NetworkProcess/NetworkDataTask.cpp:
(WebKit::NetworkDataTask::create):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.cpp:
(WebKit::NetworkDataTaskCurl::NetworkDataTaskCurl):
(WebKit::NetworkDataTaskCurl::curlDidReceiveResponse):
(WebKit::NetworkDataTaskCurl::curlDidComplete):
(WebKit::NetworkDataTaskCurl::updateNetworkLoadMetrics):
* Source/WebKit/NetworkProcess/curl/NetworkDataTaskCurl.h:

Canonical link: https://commits.webkit.org/256044@main
